From d230ea3e20a0c12e34cfe2999254ee38d8dd94fe Mon Sep 17 00:00:00 2001
From: Myron Stowe <myron.stowe@redhat.com>
Date: Wed, 6 Aug 2014 03:47:30 -0400
Subject: [pci] pciehp: Don't turn slot off when hot-added device already exists

Message-id: <20140806034730.1438.62073.stgit@gir.stowe>
Patchwork-id: 87441
O-Subject: [RHEL7.1 PATCH 32/69] PCI: pciehp: Don't turn slot off when hot-added device already exists
Bugzilla: 1110896
RH-Acked-by: Stefan Assmann <sassmann@redhat.com>
RH-Acked-by: Dean Nelson <dnelson@redhat.com>
RH-Acked-by: Prarit Bhargava <prarit@redhat.com>
RH-Acked-by: Tony Camuso <tcamuso@redhat.com>

Bugzilla: https://bugzilla.redhat.com/show_bug.cgi?id=1110896
Upstream Status: 50277c8b06d56f2345e1a0693db46db29fc6d063

commit 50277c8b06d56f2345e1a0693db46db29fc6d063
Author: Yijing Wang <wangyijing@huawei.com>
Date:   Tue Feb 11 17:36:51 2014 -0700

    PCI: pciehp: Don't turn slot off when hot-added device already exists

    If we found device already exists during hot add device, we should leave
    it, not turn the slot off.

    Signed-off-by: Yijing Wang <wangyijing@huawei.com>
    Signed-off-by: Bjorn Helgaas <bhelgaas@google.com>

Signed-off-by: Jarod Wilson <jarod@redhat.com>

diff --git a/drivers/pci/hotplug/pciehp_ctrl.c b/drivers/pci/hotplug/pciehp_ctrl.c
index fec99a1..b4a4ac1 100644
--- a/drivers/pci/hotplug/pciehp_ctrl.c
+++ b/drivers/pci/hotplug/pciehp_ctrl.c
@@ -233,7 +233,8 @@ static int board_added(struct slot *p_slot)
  if (retval) {
   ctrl_err(ctrl, "Cannot add device at %04x:%02x:00\n",
     pci_domain_nr(parent), parent->number);
-  goto err_exit;
+  if (retval != -EEXIST)
+   goto err_exit;
  }
 
  pciehp_green_led_on(p_slot);
diff --git a/drivers/pci/hotplug/pciehp_pci.c b/drivers/pci/hotplug/pciehp_pci.c
index b07d7cc..1b53306 100644
--- a/drivers/pci/hotplug/pciehp_pci.c
+++ b/drivers/pci/hotplug/pciehp_pci.c
@@ -50,7 +50,7 @@ int pciehp_configure_device(struct slot *p_slot)
     "at %04x:%02x:00, cannot hot-add\n", pci_name(dev),
     pci_domain_nr(parent), parent->number);
   pci_dev_put(dev);
-  ret = -EINVAL;
+  ret = -EEXIST;
   goto out;
  }
 
-- 
1.7.1